File: 백준/Gold/16236. 아기 상어/아기 상어.cc
#include <vector>
#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <cstring>
#include <queue>
#include <cmath>

using namespace std;

struct p {
    int x, y, d;
};

int dist(const p& a, const p& b) {
    return abs(a.x - b.x) + abs(a.y - b.y);
}

bool operator<(const p& a, const p& b) {
    if (a.d != b.d) return a.d < b.d;
    return a.x != b.x ? a.x < b.x : a.y < b.y;
}

int n;
int sea[20][20];
int visit[20][20];

vector<p> dv = {{0, -1}, {0, 1}, {1, 0}, {-1, 0}};
p q[400];

p cur;
p cand;
int big = 2, ate;
constexpr int never_reach = 999;

void bfs(int lv) {
    int s = 0, e = 0;

    cur.d = 0;
    q[e++] = cur;
    visit[cur.x][cur.y]++;

    while (s < e) {
        p now = q[s++];
        if (now.d >= cand.d) break;
        for (auto d : dv) {
            p nxt = {now.x + d.x, now.y + d.y, now.d + 1};
            if (nxt.x < 0 || nxt.x >= n || nxt.y < 0 || nxt.y >= n || sea[nxt.x][nxt.y] > big) continue;
            if (visit[nxt.x][nxt.y] != lv) {
                visit[nxt.x][nxt.y] = lv;
                q[e++] = nxt;
                if (sea[nxt.x][nxt.y] != 0 && sea[nxt.x][nxt.y] < big && nxt < cand) cand = nxt;
            }
        }
    }
}

int main () {
    scanf("%d", &n);

    int x;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            scanf("%d", &x);
            if (x == 9) {
                cur = {i, j};
                sea[i][j] = 0;
            } else {
                sea[i][j] = x;
            }
        }
    }
    
    int ans = 0;
    int lv = 1;
    while (true) {
        cand = {99, 99, never_reach};
        bfs(lv++);
        if (cand.d == never_reach) break;
        sea[cand.x][cand.y] = 0;
        ate++;
        if (ate == big) {
            big++;
            ate = 0;
        }
        ans += cand.d;
        cur = cand;
    }

    printf("%d\n", ans);
}

File: 백준/Gold/16236. 아기 상어/아기 상어.rs
use std::io::{stdin, Read, BufReader};
use std::fs::File;
use std::collections::{VecDeque};
use std::cmp::*;

#[derive(Clone, Copy, PartialEq, Eq)]
struct P {
    x: usize, y: usize, d: usize,
}

impl PartialOrd for P {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        if self.d != other.d {
            return self.d.partial_cmp(&other.d);
        }
        if self.x != other.x {
            return self.x.partial_cmp(&other.x);
        }
        self.y.partial_cmp(&other.y)
    }
}

impl Ord for P {
    fn cmp(&self, other: &Self) -> Ordering {
        self.partial_cmp(other).unwrap()
    }
}

struct Sea {
    n: isize,
    sea: [[usize; 20]; 20],
    visit: [[i32; 20]; 20],
}

const NEVER_REACH: usize = 999;

fn bfs(big: usize, mut cur: P, map: &mut Sea, lv: i32) -> Option<P> {
    let dv = [(0, -1), (0, 1), (1, 0), (-1, 0)];

    let mut food = P {x: 999, y: 999, d: NEVER_REACH};
    let mut q = VecDeque::new();

    cur.d = 0;
    map.visit[cur.x][cur.y] = lv;
    q.push_back(cur);
    
    while let Some(now) = q.pop_front() {
        if now.d >= food.d {
            break;
        }
        for &(dx, dy) in dv.iter() {
            let (nx, ny) = (now.x as isize + dx, now.y as isize + dy);
            if !(0..map.n).contains(&nx) || !(0..map.n).contains(&ny) {
                continue;
            }

            let nxt = P {x: nx as usize, y: ny as usize, d: now.d + 1};
            if map.sea[nxt.x][nxt.y] > big || map.visit[nxt.x][nxt.y] == lv {
                continue;
            }

            if map.sea[nxt.x][nxt.y] != 0 && map.sea[nxt.x][nxt.y] < big && nxt < food {
                food = nxt;
            }
            map.visit[nxt.x][nxt.y] = lv;
            q.push_back(nxt);
        }
    }

    if food.d == NEVER_REACH { None } else { Some(food) }
}

fn main() {
    let mut reader = BufReader::new(stdin());
    
    let mut input = String::new();
    reader.read_to_string(&mut input).unwrap();
    let mut input = input.split_ascii_whitespace().flat_map(str::parse::<usize>);

    let n = input.next().unwrap();
    let mut map = Sea{n: n as isize, sea: [[0; 20]; 20], visit: [[0; 20]; 20]};

    let mut ans = 0;
    let (mut size, mut ate) = (2, 0);
    let mut cur = P{x: 0, y: 0, d: 0};
    let mut lv = 1;

    for i in 0..n {
        for j in 0..n {
            let mut x = input.next().unwrap();
            if x == 9 {
                cur = P{x: i, y: j, d: 0};
                x = 0;
            }
            map.sea[i][j] = x;
        }
    }

    while let Some(food) = bfs(size, cur, &mut map, lv) {
        lv += 1;
        ate += 1;
        map.sea[food.x][food.y] = 0;
        if size == ate {
            size += 1;
            ate = 0;
        }
        ans += food.d;
        cur = food;
    }

    println!("{}", ans);
}
